File: LeetCode/Medium/Spiral Matrix/Java/spiralMatrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int rowSize = matrix.length;
        int colSize = matrix[0].length;
        int left = 0;
        int right = colSize-1;
        int up = 0;
        int down = rowSize-1;
        List<Integer> spiralArr = new ArrayList<>();

        while(spiralArr.size() < rowSize * colSize) {
            for (int j=left; j<=right; j++) spiralArr.add(matrix[up][j]);

            for (int i=up+1; i<=down; i++) spiralArr.add(matrix[i][right]);

            if (up != down) 
                for (int j=right-1; j>=left; j--) spiralArr.add(matrix[down][j]);

            if (left != right)
                for (int i=down-1; i>up; i--) spiralArr.add(matrix[i][left]);
            
            left++;
            right--;
            up++;
            down--;
        }

        return spiralArr;
    }
}

File: LeetCode/Medium/Spiral Matrix/Go/spiralMatrix.go
package main

import "fmt"

func spiralOrder(matrix [][]int) []int {
	rowSize := len(matrix)
	colSize := len(matrix[0])
	up := 0
	down := rowSize - 1
	left := 0
	right := colSize - 1

	spiralArr := []int{}
	for len(spiralArr) < rowSize*colSize {
		for j := left; j <= right; j++ {
			spiralArr = append(spiralArr, matrix[up][j])
		}

		for i := up + 1; i <= down; i++ {
			spiralArr = append(spiralArr, matrix[i][right])
		}

		if up != down {
			for j := right - 1; j >= left; j-- {
				spiralArr = append(spiralArr, matrix[down][j])
			}
		}

		if left != right {
			for i := down - 1; i > up; i-- {
				spiralArr = append(spiralArr, matrix[i][left])
			}
		}

		up++
		down--
		left++
		right--
	}
	return spiralArr
}

func main() {
	fmt.Println(spiralOrder([][]int{
		{1, 2, 3}, {4, 5, 6}, {7, 8, 9},
	}))
}
